tests/global_communication/test_communicator.cpp

```diff
@@ -14,22 +14,126 @@ using namespace nest::mc;
+
+    const auto num_domains = policy::size();
+    const auto rank = policy::id();
+
+    // dry run mode is a special case
+    if (is_dry_run()) {
+        const auto n_local_spikes = 10;
+        const auto n_local_cells = n_local_spikes;
 
-    auto part = util::parition_view(counts);
-    for(auto p : part) {
-        EXPECT_EQ(p.second-p.first, 1);
+        // Important: set up meta-data in dry run back end.
+        policy::set_sizes(policy::size(), n_local_cells);
+
+        // create local spikes for communication
+        std::vector<spike_proxy> local_spikes;
+        for (auto i=0; i<n_local_spikes; ++i) {
+            local_spikes.push_back(spike_proxy{i, rank});
+        }
+
+        // perform exchange
+        const auto global_spikes = policy::gather_spikes(local_spikes);
+
+        // test that partition information is correct
+        const auto& part = global_spikes.partition();
+        EXPECT_EQ(num_domains+1u, part.size());
+        for (auto i=0u; i<part.size(); ++i) {
+            EXPECT_EQ(part[i], n_local_spikes*i);
+        }
+
+        // test that spikes were correctly exchanged
+        //
+        // The local spikes had sources numbered 0:n_local_spikes-1
+        // The global exchange should replicate the local spikes and
+        // shift their sources to make them local to the "dummy" source
+        // domain.
+        // We set the model up with n_local_cells==n_local_spikes with
+        // one spike per local cell, so the result of the global exchange
+        // is a list of num_domains*n_local_spikes spikes that have
+        // contiguous source gid
+        const auto& spikes = global_spikes.values();
+        for (auto i=0u; i<spikes.size(); ++i) {
+            const auto s = spikes[i];
+            EXPECT_EQ(i, unsigned(s.source.gid));
+            EXPECT_EQ(0, s.domain);
+        }
+    }
+    else {
+        const auto scale = 10;
+        auto sumn = [scale](int n) {return scale*n*(n+1)/2;};
+        const auto n_local_spikes = scale*rank;
+
+        // create local spikes for communication
+        // the ranks generate different numbers of spikes, with the ranks
+        // generating the following number of spikes
+        //      [ 0, scale, 2*scale, 3*scale, ..., (num_domains-1)*scale ]
+        // i.e. 0 spikes on the first rank, scale spikes on the second, and so on.
+        std::vector<spike_proxy> local_spikes;
+        const auto local_start_id = sumn(rank-1);
+        for (auto i=0; i<n_local_spikes; ++i) {
+            local_spikes.push_back(spike_proxy{local_start_id+i, rank});
+        }
+
+        // perform exchange
+        const auto global_spikes = policy::gather_spikes(local_spikes);
+
+        // test that partition information is correct
+        const auto& part =global_spikes.partition();
+        EXPECT_EQ(unsigned(num_domains+1), part.size());
+        EXPECT_EQ(0, (int)part[0]);
+        for (auto i=1u; i<part.size(); ++i) {
```

Comment:
Because partition maps are [0, length1, length2, length3] so that you can also treat them as bound pairs [0, length1), [length1, length2)... The other choice is to go from zero, but have the EXPECT_EQ(sumn(i), (int) part[i+1]).